### DISCUSSION TITLE
mpi4py.util.pool - in analogy to multiprocessing.pool

### DISCUSSION THREAD

**Discussion by jan-janssen** [mpi4py/mpi4py#318] [Ideas]
The standard library in python for parallel processes on a single node is `multiprocessing`[1], so it would be great to provide a similar interface for `mpi4py`. I created a first draft of such an interface `pympipool`[2]. The syntax mimics the syntax of the `multiprocessing.pool`: 
```
import numpy as np
from pympipool import Pool

def calc(i):
    return np.array(i ** 2)

with Pool(cores=2) as p:
    print(p.map(function=calc, lst=[1, 2, 3, 4]))
```
The script can be executed using a simple `python test.py` command, still in the background (in a subprocess) it is using the `mpi4py.futures.MPIPoolExecutor` to spawn the subprocesses and aggregate the results. The communication between the serial python process and the mpi parallel python process is handled via `zmq`[3]. An initial discussion on this topic can be found in https://github.com/mpi4py/mpi4py/issues/232 .

[1]: https://docs.python.org/3/library/multiprocessing.html
[2]: https://github.com/pyiron/pympipool
[3]: https://zeromq.org

**Comment by dalcinl** [comment]
Tentative implementation in this branch: https://github.com/mpi4py/mpi4py/tree/feature/util-pool

**Reply by dalcinl** [reply]
@leofang I believe the multiprocesing API is ugly and dated. Even core Python docs recommend to look at `concurrent.futures` (*See also:* admonition in the [introduction](https://docs.python.org/3/library/multiprocessing.html#introduction)). However, @jan-janssen made the point that many folks are used to that API, and that adding something like this to mpi4py may be useful for long time users of multiprocessing. The only drawback I see in adding this, besides the potential future maintenance work, is in somehow advocating and keep supporting the use of an outdated model. But who I am to judge?. What do you think?

**Reply by leofang** [reply]
Sorry for my late reply, @dalcinl, awfully bad at catching up with OSS developments...

I didn't know until now (or maybe I knew but forgot) that `MPIPoolExecutor` has supported `starmap()`. This is why you were able to make `mpi4py.util.pool.Pool` a thin wrapper on top of it (docs/typing aside). `submit()` + `map()` + `starmap()` allowed you to implement all public `Pool` APIs. I think it's a great approach. 

Ofc the maintenance burden would be on you as the main maintainer (I am always happy to help, but I realize I could be trapped in another release hell like the one I just escaped from), so I obviously cannot speak for you, but I also doubt there's much work to do once after the first one or two iterations (such as addressing bug reports). 

Advocation is an interesting aspect. I don't think this time it's worth a paper, but I also don't know if we care? Maybe an announcement on GH + mailing list is enough these days? I don't know 🤷‍♂️

**Reply by dalcinl** [reply]
Our "competitor" Ray has a `Pool` implementation [[docs]](https://docs.ray.io/en/latest/ray-more-libs/multiprocessing.html).
I'll follow @jan-janssen original rationale about user convenience and expose this new utility module in mpi4py.

**Comment by dalcinl** [comment]
See #340.

**Comment by leofang** [comment]
(Let me reopen this discussion and take an answer from Lisandro. I noticed that if a discussion is marked as closed, at the entry time to GitHub Discussions it's not visible to readers, unless they specifically look for "closed" or "all" discussions.)

**Reply by leofang** [reply]
OK, this one is in the "Ideas" category, not "Q&A", and we can't mark any reply as the answer...

**Reply by dalcinl** [reply]
I closed this "idea" simply because it was implemented. You can move this discussion to the Q&A section, but I'm not really sure what do you plan to mark as "the answer".
